### PR TITLE
chapter_10

### PR DESCRIPTION
**Pull request by FirdausIsa**
ready

### CODE REVIEW

**Comment by FirdausIsa** [comment]
ready

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,5 +1,12 @@
 class UsersController < ApplicationController
   before_action :load_user, only: :show
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:3]: except: %i(new show create)

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,5 +1,12 @@
...
+  before_action :correct_user, only: [:edit, :update]
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:4]:
```suggestion
  before_action :correct_user, only: %i(edit update)
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -18,7 +25,29 @@ def create
...
+    @user = User.find_by id: params[:id]
+  end
+
+  def update
+    @user = User.find_by(id: params[:id])
+    if @user.update(user_params)
+      flash[:success] = t(:profile_updated)
+      redirect_to @user
+    else
+      render "edit"
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:38]:
```suggestion
      render :edit
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -31,4 +60,25 @@ def load_user
...
+    return if current_user?(@user)
+    flash[:danger] = t(:n_authorized)
+    redirect_to(root_url) unless current_user?(@user)
+  end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
+  end
+   
+  private
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:83]: Private for what?

File: app/models/user.rb [hunk 5]
```diff
@@ -37,4 +37,10 @@ def remember
   def forget
     update_attribute :remember_digest, nil
   end
+
+  def authenticated? remember_token
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+  
+  private
```
Comment (anlvq-1058) [inline app/models/user.rb:45]: let the method name downcase_email under private

File: app/views/users/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  <%= f.label :name, t(:name) %>
+  <%= f.text_field :name, class: "form-control" %>
+  <%= f.label :email, t(:email) %>
+  <%= f.email_field :email, class: "form-control" %>
+  <%= f.label :password, t(:password) %>
+  <%= f.password_field :password, class: "form-control" %>
+  <%= f.label :password_confirmation, t(:p_confirmation) %>
+  <%= f.password_field :password_confirmation, class: "form-control" %>
+  <%= f.submit t(:save_changes), class: "btn btn-primary" %>
+<% end %>
```
Comment (anlvq-1058) [inline app/views/users/_form.html.erb:12]: endline

File: config/locales/en.yml [hunk 7]
```diff
@@ -67,3 +67,14 @@ en:
...
+  edit_user: "Edit user"
+  save_changes: "Save changes"
+  update_profile: "Update your profile"
+  p_confirmation: "Password Confirmation"
+  pls_log_in: "Please log in."
+  profile_updated: "Profile updated"
+  user_deleted: "User deleted"
+  delete_fail: "Delete fail!"
+  n_authorized: "You are not authorized"
+
```
Comment (anlvq-1058) [inline config/locales/en.yml:80]: last line have space please remove the space

File: config/locales/ja.yml [hunk 8]
```diff
@@ -36,4 +36,14 @@ ja:
...
+  edit_user: "ユーザーの編集"
+  save_changes: "変更内容を保存"
+  update_profile: "プロフィールを更新する"
+  p_confirmation: "パスワードの確認"
+  pls_log_in: "ログインしてください。"
+  profile_updated: "プロフィール更新"
+  user_deleted: "ユーザーが削除されました"
+  delete_fail: "削除失敗！"
+  n_authorized: "許可されていません"
+
```
Comment (anlvq-1058) [inline config/locales/ja.yml:49]: last line have space please remove the space

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by FirdausIsa** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**